draft-ietf-tls-esni.md

```diff
@@ -648,46 +647,27 @@ may actually be invalid for one or the other ClientHello, in which case a fresh
 ClientHello MUST be generated, ignoring the instructions in HelloRetryRequest.
 Otherwise, the usual rules for HelloRetryRequest processing apply.
 
-Clients bind encryption of the second ClientHelloInner to encryption of the
-first ClientHelloInner via the derived ech_hrr_key by modifying HPKE setup as
-follows:
+The client encodes the second ClientHelloInner as in {{encoding-inner}}, using
+the second ClientHelloOuter for any referenced extensions. It then encrypts
+the new EncodedClientHelloInner value as a second message with the previous
+HPKE context:
 
 ~~~
-    pkR = Deserialize(ECHConfig.public_key)
-    enc, context = SetupPSKS(pkR, "tls ech" || 0x00 || ECHConfig,
-                             ech_hrr_key, "hrr key")
-~~~
-
-The `info` parameter to SetupPSKS is the concatenation of "tls ech", a
-zero byte, and the serialized ECHConfig. Clients then encrypt the second
-ClientHelloInner using this new HPKE context. In doing so, the encrypted value
-is also authenticated by ech_hrr_key. The rationale for this is described in
-{{flow-hrr-hijack}}.
-
-Client-facing servers perform the corresponding process when decrypting second
-ClientHelloInner messages. In particular, upon receipt of a second ClientHello
-message with a ClientECH value, servers set up their HPKE context and
-decrypt ClientECH as follows:
-
-~~~
-    context = SetupPSKR(ClientECH.enc, skR,
-        "tls ech" || 0x00 || ECHConfig, ech_hrr_key, "hrr key")
-    EncodedClientHelloInner = context.Open(ClientHelloOuterAAD,
-                                           ClientECH.payload)
+    payload = context.Seal(ClientHelloOuterAAD,
+                           EncodedClientHelloInner)
 ~~~
 
-ClientHelloOuterAAD is computed from the second ClientHelloOuter as described
-in {{authenticating-outer}}. The `info` parameter to SetupPSKR is computed as
-above.
+ClientHelloOuterAAD is computed as described in {{authenticating-outer}}, but
+again using the second ClientHelloOuter. The `cipher_suite`, `config_id`, and
+`enc` fields in the new ClientECH structure are copied from the first
```

Comment:
This fields also don't matter. I wonder if we should say you empty them instead...

(Insert usual remark that I think we got HelloRetryRequest wrong. :-) )

Reply:
They need to be present in order for stateless-HRR servers to correctly re-compute the HPKE context, right? Incidentally, how do stateless-HRR servers correctly compute the nonce?

Reply:
Stateless-HRR servers don't work at all right now anyway (#333), so we can't really talk about what goes in the cookie because there's no way for the client-facing server to contribute to the cookie right now. When we do, yes, we should say something about this. (See also https://github.com/tlswg/draft-ietf-tls-esni/issues/349#issuecomment-713724970)

As to what the stateless-HRR server needs, it needs to store *something* that allows it to recreate the HPKE context on the second ClientHello. Things that would work:

* Store the HPKE encryption key, nonce, and sequence number. (Alternatively, know the sequence number is always 1...)
* Store the client `enc` value and sequence number (or assume 1), and some handle to the ECHConfig private key, so it can recreate the HPKE context and fast-forward the context to the right sequence number.
* Store a hash of the client `enc` value, check it matches, and then do as above.

Only the last of these actually needs the client to send `enc` again. Note what would not work great is blindly recomputing the `enc` value, given {{flow-hrr-hijack}}. Given #325, my inclination is the first option is most straightforward anyway.

Reply:
> Stateless-HRR servers don't work at all right now anyway (#333), so we can't really talk about what goes in the cookie because there's no way for the client-facing server to contribute to the cookie right now.

Isn't this only true for Split Mode / stateless-HRR?

Reply:
It's true for both. There's no text supporting Shared Mode / stateless-HRR right now because Shared Mode is treated as a special-case of Split Mode. It's "obvious" what they're supposed to do, but without any text to support it, we can't talk about what goes in the cookie.

As part of #333, we'll need to address all of that. In the meantime, this PR doesn't change things because we already have no text talking about how ech_hrr_key goes in the cookie. All of that is #333.

Reply:
Ack. You can resolve this comment.

Reply:
Unresolving this since we only resolved @cjpatton's clarification question, not the actual issue.

Thinking about this some more, I think we actually should require `enc` be empty on CH2. The reason is, of the three stateless implementation options listed above, only the first one is robust. The second and third rely on the server performing a check that, if they forget, still interops. But if they forget the check, {{flow-hrr-hijack}} applies again. Given @cjpatton's question, I also suspect this mistake is reasonably likely.

Clearing `enc` forces servers go pick the first option, which is the more robust one. We're also guaranteed that a server, stateful or stateless, does not need `enc` (or `cipher_suite`) because you need to process ECH on CH1 regardless. Thus, I propose:

* CH2's `enc` MUST be empty. Servers check this when parsing (since it's easy).
* CH2's  `cipher_suite` MUST be {copied over, zero; not sure which}. Servers check this one too, just because it's really easy.

Even more minimal would be if CH2's extension contained `payload` and nothing else, but we don't have any precedent for extension syntaxes varying between CH1 and CH2, so probably we shouldn't go down that route.

@ekr, I'm curious what your thoughts on this are.

Reply:
Oh same probably goes for `config_id` in CH2. Force the server to use the same ECHConfig rather than accidentally swapping it.

Reply:
(Updated the PR with this version.)